### PR TITLE
Add type annotations

### LINKED ISSUE #10: Add type annotations
Python supports type annotations now so we might as well use them. Some info on typing from the Python folks themselves can be found [here](https://docs.python.org/3/library/typing.html).

### PR DESCRIPTION
**Pull request by pie-tras**
Add type annotations.

### CODE REVIEW

**Review by sapols — CHANGES_REQUESTED** [review]
This isn't ready to review yet. You added the return types, which is a good start, but the intent of the issue is to add type annotations everywhere. Also, I suspect there will be merge conflicts once you merge the docstrings PR, so you might wanna finish that before you start this.

Just to offer a head start, I asked GPT-4 to add the type annotations and this is what it came back with (I haven't checked this at all, so don't blindly copy it):

```python
import numpy as np
import pandas as pd
import requests
import urllib.parse
from typing import Dict, List, Union, Optional


def __datasetWillUseVersion3(baseUrl: str, dataset: str, preferVersion2: bool) -> bool:
    ...

def data(baseUrl: str, dataset: str, returnType: str, 
         projections: Optional[List[str]] = [], 
         selections: Optional[List[str]] = [], 
         operations: Optional[List[str]] = [], preferVersion2: bool = False) -> Union[np.ndarray, pd.DataFrame, None]:
    ...

def download(baseUrl: str, dataset: str, filename: str, fileFormat: str, 
             projections: Optional[List[str]] = [], 
             selections: Optional[List[str]] = [], 
             operations: Optional[List[str]] = [], preferVersion2: bool = False) -> None:
    ...

class LatisInstance:
    """A class that represents a Latis instance."""
    baseUrl: str
    latis3: bool
    catalog: "Catalog"

    def __init__(self, baseUrl: str, latis3: bool) -> None:
        ...
    
    def getDataset(self, name: str, projections: Optional[List[str]] = [],
                   selections: Optional[List[str]] = [], 
                   operations: Optional[List[str]] = []) -> "Dataset":
        ...
    
    def __formatBaseUrl(self) -> None:
        ...

    def __getCatalog(self) -> "Catalog":
        ...

class Catalog:
    """Catalog for a Latis instance."""
    datasets: Dict[str, str]
    list: np.ndarray

    def __init__(self, latisInstance: LatisInstance) -> None:
        ...

    def search(self, searchTerm: str) -> List[str]:
        ...

class Dataset:
    """A class that represents a dataset from a Latis instance."""
    latisInstance: LatisInstance
    name: str
    projections: List[str]
    selections: List[str]
    operations: List[str]
    query: Optional[str]
    metadata: "Metadata"

    def __init__(self, latisInstance: LatisInstance, name: str, 
                 projections: Optional[List[str]] = [], 
                 selections: Optional[List[str]] = [], 
                 operations: Optional[List[str]] = []) -> None:
        ...

    def select(self, target: str = "time", start: str = "", 
               end: str = "", inclusive: bool = True) -> "Dataset":
        ...

    def project(self, projectionList: List[str]) -> "Dataset":
        ...

    def operate(self, operation: str) -> "Dataset":
        ...

    def buildQuery(self) -> str:
        ...

    def asPandas(self) -> pd.DataFrame:
        ...

    def asNumpy(self) -> np.ndarray:
        ...

    def getFile(self, filename: str, format: str = 'csv') -> None:
        ...

    def clearProjections(self) -> None:
        ...

    def clearSelections(self) -> None:
        ...

    def clearOperations(self) -> None:
        ...

class Metadata:
    """A class that represents metadata for a dataset."""
    properties: Dict[str, Union[str, int, float, bool]]

    def __init__(self, latisInstance: LatisInstance, dataset: Dataset) -> None:
        ...

```

**Comment by lindholc** [comment]
I think we'll want to add a pinned version of `mypy` to the dev dependencies so developers will have a consistent version to use for checking.

I just installed the latest `mypy` on 3.11 (I don't know which versions we're targeting) to test and ran into some issues.

First, `latis-python-client` isn't a valid Python module name, so the directory containing `__init__.py` probably needs to be renamed. `mypy` chokes on this, but you also won't be able to import the client to use it. I think with a tree like the following, you'd be able to write `import latis.client` or `from latis import client`:

```
src
└── latis
    ├── __init__.py
    └── client.py
```

After tweaking the directory name to be a valid module name, I got the following errors from `mypy`:

```
src/latis_python_client/latis.py:9: error: Library stubs not installed for "pandas"  [import]
src/latis_python_client/latis.py:9: note: Hint: "python3 -m pip install pandas-stubs"
src/latis_python_client/latis.py:9: note: (or run "mypy --install-types" to install all missing stub packages)
src/latis_python_client/latis.py:9: note: See https://mypy.readthedocs.io/en/stable/running_mypy.html#missing-imports
src/latis_python_client/latis.py:10: error: Library stubs not installed for "requests"  [import]
src/latis_python_client/latis.py:10: note: Hint: "python3 -m pip install types-requests"
src/latis_python_client/latis.py:263: error: Argument 1 to "list" has incompatible type "Optional[List[str]]"; expected "Iterable[str]"  [arg-type]
src/latis_python_client/latis.py:264: error: Argument 1 to "list" has incompatible type "Optional[List[str]]"; expected "Iterable[str]"  [arg-type]
src/latis_python_client/latis.py:265: error: Argument 1 to "list" has incompatible type "Optional[List[str]]"; expected "Iterable[str]"  [arg-type]
src/latis_python_client/latis.py:340: error: Incompatible types in assignment (expression has type "str", variable has type "None")  [assignment]
src/latis_python_client/latis.py:342: error: Unsupported left operand type for + ("None")  [operator]
src/latis_python_client/latis.py:345: error: Unsupported left operand type for + ("None")  [operator]
src/latis_python_client/latis.py:348: error: Unsupported left operand type for + ("None")  [operator]
src/latis_python_client/latis.py:350: error: Incompatible return value type (got "None", expected "str")  [return-value]
src/latis_python_client/latis.py:363: error: Name "numpy.DataFrame" is not defined  [name-defined]
src/latis_python_client/latis.py:391: error: "None" has no attribute "replace"  [attr-defined]
```

I think the missing stubs would also belong in the dev dependencies.

File: requirements.txt [hunk 1]
```diff
@@ -1,3 +1,4 @@
+mypy==3.10
```
Comment (pie-tras) [inline requirements.txt:1]: Resolved mypy errors and added an install requirement for version 3.10

File: tests/example.py [hunk 2]
```diff
@@ -6,14 +6,14 @@
...
+                              '\\src\\latispy')
 else:
     libPath = libPath.replace('/tests',
-                              '/src/latis-python-client')
+                              '/src/latispy')
 
 sys.path.insert(0, libPath)
 
-import latis
+import latispy as latis
```
Comment (pie-tras) [inline tests/example.py:16]: latispy may be a bit redundant? Perhaps just latis would be a better valid module name?

**Review by sapols — APPROVED** [review]
LGTM 👍

**Pull request merged**